lib/trie/trie.go

```diff
@@ -1157,8 +1383,32 @@ func handleDeletion(branch *Node, key []byte) (newNode *Node, branchChildMerged
 			}
 		}
 
-		return newBranch, branchChildMerged
+		return newBranch, branchChildMerged, nil
+	}
+}
+
+// ensureMerkleValueIsCalculated is used before calling PopulateMerkleValues
+// to ensure the parent node and all its descendant nodes have their Merkle
+// value computed and ready to be used. This has a close to zero performance
+// impact if the parent node Merkle value is already computed.
+func ensureMerkleValueIsCalculated(parent *Node, isRoot bool) (err error) {
```

Comment:
I changed `registerDeletedMerkleValue` to be a receiver on Trie so we don't have to spam with `isRoot := t.root == parent` in the trie code.

However, I left `ensureMerkleValueIsCalculated` as is, since the next PR (branch based on this one) tracks inserted node hashes, and we need to inject `isRoot` from the caller to that function since the node to be created is never equal to the current trie root, but may replace it (so `isRoot` should be true).

Anyway, this is about to be thrown out with lazy loading since every node will have its Merkle value <-> encoding stored on disk.

Reply:
Changed `ensureMerkleValueIsCalculated` to check for the root node within its body since the PR on tracking node hashes on every insertion (#2881) got abandoned.